fedimint-client-rpc/src/lib.rs

```diff
@@ -0,0 +1,646 @@
+use tracing::info;
+
+// Key prefixes for the unified database
+#[repr(u8)]
+#[derive(Clone, Copy, Debug)]
+pub enum DbKeyPrefix {
+    ClientDatabase = 0x00,
+    Mnemonic = 0x01,
+}
+
+#[derive(Debug, Clone, Encodable, Decodable, Eq, PartialEq, Hash)]
+pub struct MnemonicKey;
+
+impl_db_record!(
+    key = MnemonicKey,
+    value = String,
+    db_prefix = DbKeyPrefix::Mnemonic,
+);
+
+/// A database wrapper that isolates client data within the unified database
+/// Each client sees only its own data by using client-specific key prefixes
+#[derive(Clone)]
+pub struct ClientDatabaseWrapper {
+    unified_db: Database,
+}
+
+impl ClientDatabaseWrapper {
+    pub fn new(unified_db: Database, client_name: String) -> Self {
+        // assert_eq!(client_name.len(), 43);
+
+        let mut client_prefix = vec![DbKeyPrefix::ClientDatabase as u8];
+        client_prefix.extend_from_slice(client_name.as_bytes());
+
+        Self {
+            unified_db: unified_db.with_prefix(client_prefix),
+        }
+    }
+}
+
+impl From<ClientDatabaseWrapper> for Database {
+    fn from(wrapper: ClientDatabaseWrapper) -> Self {
+        wrapper.unified_db
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct RpcRequest {
+    pub request_id: u64,
+    #[serde(flatten)]
+    pub kind: RpcRequestKind,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum RpcRequestKind {
+    SetMnemonic {
+        words: Vec<String>,
+    },
+    GenerateMnemonic,
+    GetMnemonic,
+    NukeData,
+    /// Join federation (requires mnemonic to be set first)
+    JoinFederation {
+        invite_code: String,
+        client_name: String,
+    },
+    OpenClient {
+        client_name: String,
+    },
+    CloseClient {
+        client_name: String,
+    },
+    ClientRpc {
+        client_name: String,
+        module: String,
+        method: String,
+        payload: serde_json::Value,
+    },
+    CancelRpc {
+        cancel_request_id: u64,
+    },
+    ParseInviteCode {
+        invite_code: String,
+    },
+    ParseBolt11Invoice {
+        invoice: String,
+    },
+    PreviewFederation {
+        invite_code: String,
+    },
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct RpcResponse {
+    pub request_id: u64,
+    #[serde(flatten)]
+    pub kind: RpcResponseKind,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum RpcResponseKind {
+    Data { data: serde_json::Value },
+    Error { error: String },
+    Aborted {},
+    End {},
+}
+
+pub trait RpcResponseHandler: MaybeSend + MaybeSync {
+    fn handle_response(&self, response: RpcResponse);
+}
+// A trait for database creation (handles both new and existing databases).
+#[apply(async_trait_maybe_send!)]
+pub trait DatabaseFactory: MaybeSend + MaybeSync + 'static {
+    async fn create_database(&self, name: &str) -> anyhow::Result<Database>;
+}
```

Comment:
you can remove this trait, just add `Database` and argument to RpcGlobalState::new